backend/cfg/cfg_regalloc_validate.ml

```diff
@@ -0,0 +1,1073 @@
+      are not considered there. *)
+  val append_equations :
+    t ->
+    tag:'a Error.Tag.t ->
+    exn:t option ->
+    reg_instr:'a Instruction.t ->
+    loc_instr:'a instruction ->
+    destroyed:Location.t array ->
+    t
+
+  val print : Format.formatter -> t -> unit
+
+  (** Convert the domain value to its representation. *)
+  val to_result : t -> (Equation_set.t, Error.t) Result.t
+end = struct
+  module Error = struct
+    type 'a unpacked =
+      { equations : Equation_set.t;
+        exn_equations : Equation_set.t option;
+        reg_instr : 'a Instruction.t;
+        loc_instr : 'a instruction;
+        message : string
+      }
+
+    module Tag = struct
+      type 'a t =
+        | Terminator : terminator t
+        | Basic : basic t
+    end
+
+    type t =
+      | Terminator : terminator unpacked -> t
+      | Basic : basic unpacked -> t
+
+    let print ppf (error : t) =
+      let message, equations, exn_equations, id, reg_instr, loc_instr =
+        match error with
+        | Basic { loc_instr; reg_instr; equations; exn_equations; message } ->
+          ( message,
+            equations,
+            exn_equations,
+            loc_instr.id,
+            `Basic (Instruction.to_prealloc ~alloced:loc_instr reg_instr),
+            `Basic loc_instr )
+        | Terminator { loc_instr; reg_instr; equations; exn_equations; message }
+          ->
+          ( message,
+            equations,
+            exn_equations,
+            loc_instr.id,
+            `Terminator (Instruction.to_prealloc ~alloced:loc_instr reg_instr),
+            `Terminator loc_instr )
+      in
+      Format.fprintf ppf "CFG REGALLOC Check failed in instr %d:\n" id;
+      Format.fprintf ppf "Instruction's description before allocation: %a\n"
+        Cfg.print_instruction reg_instr;
+      Format.fprintf ppf "Instruction's description after allocation: %a\n"
+        (Cfg.print_instruction' ~print_reg:print_reg_as_loc)
+        loc_instr;
+      Format.fprintf ppf "Message: %s\n" message;
+      Format.fprintf ppf "Live equations for the normal successor: [%a]\n"
+        Equation_set.print equations;
+      Option.iter
+        (fun exn_equations ->
+          Format.fprintf ppf
+            "Live equations for the exceptional successor: [%a]\n"
+            Equation_set.print exn_equations)
+        exn_equations;
+      ()
+  end
+
+  (** This type logically works as [(Equation_set.t, Error.t) Result.t] but in
+      case there's an error we want to know the "last known equations" for a
+      given instruction. Therefore [{equations; error = None}] corresponds to a
+      [Ok equations] and [{equations; error = Some error}] corresponds to [Error
+      error] where [equations] is biggest set of equations for a given
+      instruction known before the error was found. All the values inside
+      [error] correspond to values from the point of error. *)
+  type t =
+    { equations : Equation_set.t;
+      error : Error.t option
+    }
```

Comment:
Is this complexity in representation necessary? Powerset is such a nice domain, can we just work with it? Why do we need to propagate anything after we encounter an error? The transformers can just raise, and the data flow analysis stop and return the currently available abstract values.

Reply:
Because we want to be able to debug the errors.
This gives us Cfg with all the partial equations annotated so the source of the equation (and be implication source of the error) can be found.

Reply:
Changed the dataflow, now the we have `type domain = Equation_set.t`